budgetkey_data_pipelines/pipelines/budget/national/changes/processed/extract-change-groups.py

```diff
@@ -0,0 +1,288 @@
+                                num_found += len(transfer_codes)
+                                to_append = {
+                                    'transfer_ids': transfer_codes,
+                                    'date': date,
+                                    'pending': pending,
+                                }
+                                result.append(to_append)
+                                found = True
+                    except StopIteration:
+                        pass
+        for change in changes:
+            if not change['trcode'] in selected_transfer_codes:
+                selected_transfer_codes.add(change['trcode'])
+                result.append({
+                    'transfer_ids': [change['trcode']],
+                    'date': change['date'],
+                    'pending': change['pending']
+                })
+        return result
+
+    def build_groups(groups, changes):
+        for group in groups:
+            trcodes = set(group['transfer_ids'])
+            years = set(
+                int(x.split('/')[0])
+                for x in trcodes
+            )
+            assert(len(years) == 1)
+
+            group['transfer_ids'] = list(set(
+                x.split('/')[1]
+                for x in trcodes
+            ))
+            group['group_id'] = group['transfer_ids'][0]
+
+            transfer_changes = list(filter(
+                lambda x: x['trcode'] in trcodes,
+                changes
+            ))
+            group['budget_codes'] = list(set(
+                x['budget_code']
+                for x in transfer_changes
+            ))
+            group['prefixes'] = list(set(
+                itertools.chain.from_iterable(
+                    [code[:l] for l in range(2, 10, 2)]
+                    for code in group['budget_codes']
+                )
+            ))
+            group['prefixes'].sort(key=lambda x: int('1' + x))
+            group['changes'] = [
+                {'keys': get_keys_for_prefix(transfer_changes, code)}
+                for code in group['prefixes']
+            ]
+
+            trcodes = list(trcodes)
+            trcodes.sort()
+            for trcode in trcodes:
+                per_transfer_changes = list(filter(
+                    lambda x:
+                        x['trcode'] == trcode and
+                        not x['budget_code'].startswith('0047'),
+                    transfer_changes
+                ))
+                s = sum(
+                    sum(x[f] for f in value_fields)
+                    for x in per_transfer_changes
+                )
+                if s > 0:
+                    group['group_id'] = trcode.split('/')[1]
+                    logging.debug(
+                        'selected group id %(group_id)s as '
+                        'representative for %(transfer_ids)r' % group)
+                    break
+
+        return groups
+
+    changes.sort(key=get_date)
```

Comment:
instead of working with the entire list of changes, we should work with the changes year by year (by using itertools.groupby).
We will have to sort the list of changes anyway, but it's limited to the amount of changes per year (so more scalable over time).